mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -0,0 +1,423 @@
+use std::cmp;
+use std::collections::{BTreeMap, HashMap};
+use std::ops::Range;
+
+use async_trait::async_trait;
+use futures::task::{Spawn, SpawnExt};
+use futures::{pin_mut, StreamExt};
+use intervaltree::IntervalTree;
+use mountpoint_s3_client::{types::ETag, ObjectClient};
+use tracing::{debug_span, error, trace, warn, Instrument};
+
+use crate::checksums::{ChecksummedBytes, IntegrityError};
+use crate::object::ObjectId;
+use crate::prefetch::part::Part;
+use crate::prefetch::part_queue::unbounded_part_queue;
+use crate::prefetch::part_queue::PartQueueProducer;
+use crate::prefetch::part_stream::{ObjectPartStream, RequestRange};
+use crate::prefetch::task::RequestTask;
+use crate::prefetch::InMemoryCache;
+use crate::prefetch::{PrefetchReadError, RawMetadata};
+
+use super::parquet_prefetch::{ColumnIndex, RangeKey, RowGroupIndex};
+use super::{MetadataRef, RawMetadataRef};
+
+#[derive(Debug)]
+pub struct ParquetPartStream<Runtime> {
+    runtime: Runtime,
+}
+
+impl<Runtime> ParquetPartStream<Runtime> {
+    pub fn new(runtime: Runtime) -> Self {
+        Self { runtime }
+    }
+}
+
+#[async_trait]
+impl<Runtime> ObjectPartStream for ParquetPartStream<Runtime>
+where
+    Runtime: Spawn,
+{
+    #[allow(clippy::too_many_arguments)]
+    fn spawn_get_object_request<Client>(
+        &self,
+        client: &Client,
+        bucket: &str,
+        key: &str,
+        if_match: ETag,
+        range: RequestRange,
+        preferred_part_size: usize,
+        in_mem_cache: InMemoryCache,
+        parsed_metadata: MetadataRef,
+        raw_metadata: RawMetadataRef,
```

Comment:
I still don't get why we need both raw and parsed metadata, do we have it documented somewhere?